### PR TITLE
[libc++] Implement P0718R2: `atomic<shared_ptr<T>>`

### PR DESCRIPTION
**Pull request by MitalAshok**
Implements [P0718R2](https://wg21.link/P0718R2).

This is quite a "dumb" implementation: It just reuses the now-deprecated functions in [[depr.util.smartptr.shared.atomic]](https://wg21.link/depr.util.smartptr.shared.atomic) on a held `std::{shared/weak}_ptr<T>` member.

[P1644R0](https://wg21.link/P1644R0) (`wait`/`notify` for `atomic<shared_ptr<T>>`) is *not* implemented. I will work on a subsequent patch, but we need `atomic<shared_ptr>` first.

I also noticed [LWG3661](https://wg21.link/LWG3661) (`constinit atomic<shared_ptr<T>> a(nullptr);`) did not apply to `atomic<weak_ptr<T>>`, but I see no reason not to support it.

I also added forward declarations for the partial specializations in `<atomic>`. This shouldn't change the correctness of any code but should give a better error (that `atomic<shared_ptr<T>>` is incomplete, not that `shared_ptr<T>` is a non-trivial type in the primary template)

### CODE REVIEW

**Comment by mordante** [comment]
Thanks for your contribution!

> Implements [P0718R2](https://wg21.link/P0718R2).
> 
> This is quite a "dumb" implementation: It just reuses the now-deprecated functions in [[depr.util.smartptr.shared.atomic]](https://wg21.link/depr.util.smartptr.shared.atomic) on a held `std::{shared/weak}_ptr<T>` member.

Can you provide information why using deprecated features is the way to go? It seems likely these functions will be removed in C++26. Per [P2863R3](https://wg21.link/p2863r3) Review Annex D for C++26 (Alisdair Meredith) 
"
6.24 Deprecated shared_ptr atomic access [depr.util.smartptr.shared.atomic]
The legacy C-style atomic API for manipulating shared pointers provided in C++11 is subtle, frequently misunderstood, and easily misused to cause data races. A type-safe replacement facility that also provides support for atomic<weak_ptr<T>> was added to C++20, so we recommend removing the legacy API at the earliest opportunity.

See paper [[P2869](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2023/p2863r3.html#ref-P2869)] for a more detailed discussion and proposed wording.
6.24.1 SG1 (concurrency) Review, 2023 June, Varna

Polled after discussion of the header compatibility concern:

Poll: Remove deprecated shared_ptr atomic access APIs from C++26, with any of the library options listed in P2689?

SF  F  N  A SA
 2  4  1  1  0

Consensus to move this paper to LEWG to resolve the header design issue, and then continue on to LWG.
"

File: libcxx/include/__memory/atomic_shared_ptr.h [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    return std::__sp_atomic_compare_exchange_strong(std::addressof(__p), std::addressof(__e), __d);
+  }
+  _LIBCPP_HIDE_FROM_ABI bool compare_exchange_strong(_Tp& __e, _Tp __d, memory_order = memory_order_seq_cst) noexcept {
+    return std::__sp_atomic_compare_exchange_strong(std::addressof(__p), std::addressof(__e), __d);
+  }
+
+  // P1644R0 not implemented
+  // void wait(_Tp old, memory_order order = memory_order::seq_cst) const noexcept;
+  // void notify_one() noexcept;
+  // void notify_all() noexcept;
```
Comment (huixie90) [inline libcxx/include/__memory/atomic_shared_ptr.h:133]: Are these implementable given that you are just storing a `shared_ptr` ?

**Review by huixie90 — CHANGES_REQUESTED** [review]
I would say we should take the opportunity to try implementing it in a lock-free way

**Comment by ldionne** [comment]
I would be OK with using this implementation strategy as a stepping stone for a lockfree implementation, which would require making this experimental. I agree with @huixie90 that we need to implement this in a lockfree way, and my understanding is that this requires more research cause there is no efficient and publicly available implementation that works on all platforms.

**Comment by ldionne** [comment]
@timuraudio Do you have thoughts about how we could implement this lockfree?

**Comment by frederick-vs-ja** [comment]
> This is quite a "dumb" implementation: It just reuses the now-deprecated functions in [[depr.util.smartptr.shared.atomic]](https://wg21.link/depr.util.smartptr.shared.atomic) on a held `std::{shared/weak}_ptr<T>` member.

These functions are removed in C++26 ([P2869R3](https://wg21.link/p2869r3)), however, such strategy would make them not removable in libc++.

**Comment by Zingam** [comment]
@MitalAshok Is there any progress on this?

**Pull request closed without merging**